lib/galaxy/model/security.py

```diff
@@ -1447,62 +1439,171 @@ def get_showable_folders(
+    def set_user_group_and_role_associations(
+        self,
+        user: User,
+        *,
+        group_ids: Optional[List[int]] = None,
+        role_ids: Optional[List[int]] = None,
+    ) -> None:
+        """
+        Set user groups and user roles, replacing current associations.
 
-    def set_entity_group_associations(self, groups=None, users=None, roles=None, delete_existing_assocs=True):
-        users = users or []
-        roles = roles or []
-        groups = groups or []
-        for group in groups:
-            if delete_existing_assocs:
-                flush_needed = False
-                for a in group.roles + group.users:
-                    self.sa_session.delete(a)
-                    flush_needed = True
-                if flush_needed:
-                    with transaction(self.sa_session):
-                        self.sa_session.commit()
-            for role in roles:
-                self.associate_components(group=group, role=role)
-            for user in users:
-                self.associate_components(group=group, user=user)
+        Associations are set only if a list of new associations is provided.
+        If the provided list is empty, existing associations will be removed.
+        If the provided value is None, existing associations will not be updated.
+        """
+        self._persist_new_model(user)
+        if group_ids is not None:
+            self._set_user_groups(user, group_ids)
+        if role_ids is not None:
+            self._set_user_roles(user, role_ids)
+        # Commit only if both user groups and user roles have been set.
+        self.sa_session.commit()
+
+    def set_group_user_and_role_associations(
+        self,
+        group: Group,
+        *,
+        user_ids: Optional[List[int]] = None,
+        role_ids: Optional[List[int]] = None,
+    ) -> None:
+        """
+        Set group users and group roles, replacing current associations.
 
-    def set_entity_role_associations(self, roles=None, users=None, groups=None, delete_existing_assocs=True):
-        users = users or []
-        roles = roles or []
-        groups = groups or []
-        for role in roles:
-            if delete_existing_assocs:
-                flush_needed = False
-                for a in role.users + role.groups:
-                    self.sa_session.delete(a)
-                    flush_needed = True
-                if flush_needed:
-                    with transaction(self.sa_session):
-                        self.sa_session.commit()
-            for user in users:
-                self.associate_components(user=user, role=role)
-            for group in groups:
-                self.associate_components(group=group, role=role)
+        Associations are set only if a list of new associations is provided.
+        If the provided list is empty, existing associations will be removed.
+        If the provided value is None, existing associations will not be updated.
```

Comment:
I don't think `user_ids` and `role_ids` are ever `None`, see https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/schema/groups.py#L65-L70 . So maybe restore the `user_ids = user_ids or []` and `role_ids = role_ids or []` and drop the `if`s?


```suggestion
        If the provided list is empty or None, existing associations will be removed.
```

Reply:
They are never `None` simply because the UI currently provides for updating two types of associations at the same time, sending the full set of associated items for each relationship, which is then processed via the same endpoint on the backend. This applies equally to users, groups, and roles. If we wanted to add a UI feature that allows to update only one relationship (for example, user groups, but not user roles), we'd still have to retrieve all associations for groups AND roles, and send them back to the API. Which is wasteful and, I think, not great for many reasons. If we distinguish between an empty list and null, we provide for two options: "delete all associations not in the list" or "don't change anything", which is impossible with the current API. Finally, this makes testing easier and more straightforward: you only specify and pass what's being tested (I've just added a commit to actually do this!). And, finally, to second [this comment](https://github.com/galaxyproject/galaxy/pull/18374#pullrequestreview-2112781953), a null argument here, in my opinion, should be interpreted as "make no change".

Reply:
The problem is this possibility is not exposed via the API, so since this will already likely break the current API it would make sense to rewrite [GroupUpdatePayload](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/schema/groups.py#L76) so that `user_ids` and `role_ids` are Optional lists with default None.

Reply:
@nsoranzo does this look right? 5d5065adab4eb30bedbcbc92347f05505c50c1e7